workflow/controller/operator.go

```diff
@@ -3074,76 +3091,70 @@ func (woc *wfOperationCtx) fetchWorkflowSpec() (*wfv1.WorkflowSpec, error) {
 	if err != nil {
 		return nil, err
 	}
-	return specHolder.GetWorkflowSpec(), nil
+	return specHolder, nil
 }
 
-func (woc *wfOperationCtx) loadExecutionSpec() (wfv1.TemplateReferenceHolder, wfv1.Arguments, error) {
-
-	executionParameters := woc.wf.Spec.Arguments
+func (woc *wfOperationCtx) retryStrategy(tmpl *wfv1.Template) *wfv1.RetryStrategy {
+	if tmpl != nil && tmpl.RetryStrategy != nil {
+		return tmpl.RetryStrategy
+	}
+	return woc.execWf.Spec.RetryStrategy
+}
 
-	if woc.wf.Spec.WorkflowTemplateRef == nil {
-		if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
-			return nil, executionParameters, fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+func (woc *wfOperationCtx) setExecWorkflow() error {
+	if woc.wf.Spec.WorkflowTemplateRef != nil {
+		err := woc.setStoredWFSpec()
+		if err != nil {
+			return err
 		}
-		// Set the WorkflowDefaults from Configmap
+		woc.execWf = &wfv1.Workflow{Spec: *woc.wf.Status.StoredWorkflowSpec.DeepCopy()}
+		woc.volumes = woc.execWf.Spec.DeepCopy().Volumes
+	} else if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
+		return fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+	} else {
 		err := woc.controller.setWorkflowDefaults(woc.wf)
 		if err != nil {
-			log.WithFields(log.Fields{"key": woc.wf.Name, "error": err}).Warn("Failed to apply default workflow values")
-			return nil, executionParameters, err
+			return err
 		}
-
-		tmplRef := &wfv1.WorkflowStep{Template: woc.wf.Spec.Entrypoint}
-		return tmplRef, executionParameters, nil
+		woc.volumes = woc.wf.Spec.DeepCopy().Volumes
 	}
+	return nil
+}
 
+func (woc *wfOperationCtx) setStoredWFSpec() error {
 	if woc.wf.Status.StoredWorkflowSpec == nil {
-		wftSpec, err := woc.fetchWorkflowSpec()
+		wfDefault := woc.controller.Config.WorkflowDefaults
+		wftHolder, err := woc.fetchWorkflowSpec()
 		if err != nil {
-			return nil, executionParameters, err
+			return err
 		}
-		woc.wf.Status.StoredWorkflowSpec = wftSpec
+		// Join WFT and WfDefault metadata to Workflow metadata.
+		wfutil.JoinWorkflowMetaData(&woc.wf.ObjectMeta, wftHolder.GetWorkflowMetadata(), &wfDefault.ObjectMeta)
+
+		// Join WF spec WFT spec and WorkflowDefault spec
+		mergedWf, err := wfutil.JoinWorkflowSpec(&woc.wf.Spec, wftHolder.GetWorkflowSpec(), &wfDefault.Spec)
```

Comment:
Is there a reason as to why we need to do `Spec` and `ObjectMeta` separately? Why can't it all be done in one funciton?

Reply:
WorkflowTemplate's and Workflow Default's Metadata needs to merge to the original workflow's metadata for searchable/filtering (labels/annotations).